File: src/main/java/com/matag/game/turn/phases/AbstractPhase.java
package com.matag.game.turn.phases;

import static com.matag.game.turn.phases.PhaseUtils.isPriorityAllowed;
import static com.matag.game.turn.phases.beginning.UntapPhase.UT;
import static com.matag.game.turn.phases.ending.CleanupPhase.CL;

import com.matag.game.turn.phases.beginning.UntapPhase;
import org.springframework.beans.factory.annotation.Autowired;

import com.matag.game.status.GameStatus;
import com.matag.game.turn.action._continue.AutocontinueChecker;

public abstract class AbstractPhase implements Phase {
  @Autowired
  private AutocontinueChecker autocontinueChecker;

  @Autowired
  private UntapPhase untapPhase;

  @Override
  public void action(GameStatus gameStatus) {}

  @Override
  public void set(GameStatus gameStatus) {
    if (!autocontinueChecker.canPerformAnyAction(gameStatus)) {
      next(gameStatus);
    }
  }

  @Override
  public void next(GameStatus gameStatus) {
    if (gameStatus.getTurn().isEnded()) {
      return;
    }

    moveNext(gameStatus);
  }

  private void moveNext(GameStatus gameStatus) {
    if (isPriorityAllowed(getName())) {
      if (gameStatus.isCurrentPlayerActive()) {
        moveToNextPlayer(gameStatus);

      } else {
        moveToNextPhase(gameStatus);
      }

    } else {
      moveToNextPhase(gameStatus);
    }
  }

  private void moveToNextPlayer(GameStatus gameStatus) {
    gameStatus.getTurn().passPriority(gameStatus);

    if (!autocontinueChecker.canPerformAnyAction(gameStatus)) {
      moveToNextPhase(gameStatus);
    }
  }

  private void moveToNextPhase(GameStatus gameStatus) {
    Phase nextPhase;

    if (getName().equals(CL)) {
      gameStatus.getTurn().setCurrentTurnPlayer(gameStatus.getNonCurrentPlayer().getName());
      gameStatus.getTurn().increaseTurnNumber();
      nextPhase = untapPhase;

    } else {
      nextPhase = getNextPhase(gameStatus);
    }

    gameStatus.getTurn().setCurrentPhase(nextPhase.getName());
    gameStatus.getTurn().setCurrentPhaseActivePlayer(gameStatus.getTurn().getCurrentTurnPlayer());
    nextPhase.action(gameStatus);

    if (!autocontinueChecker.canPerformAnyAction(gameStatus)) {
      nextPhase.next(gameStatus);
    }
  }
}

File: src/main/kotlin/com/matag/game/turn/phases/AbstractPhase.kt
package com.matag.game.turn.phases

import com.matag.game.status.GameStatus
import com.matag.game.turn.action._continue.AutocontinueChecker
import com.matag.game.turn.phases.ending.CleanupPhase

abstract class AbstractPhase(
    private val autocontinueChecker: AutocontinueChecker? = null
) : Phase {

    override fun action(gameStatus: GameStatus) {}

    override fun set(gameStatus: GameStatus) {
        if (!autocontinueChecker!!.canPerformAnyAction(gameStatus)) {
            next(gameStatus)
        }
    }

    override fun next(gameStatus: GameStatus) {
        if (gameStatus.turn.isEnded()) {
            return
        }

        moveNext(gameStatus)
    }

    private fun moveNext(gameStatus: GameStatus) {
        if (PhaseUtils.isPriorityAllowed(name)) {
            if (gameStatus.isCurrentPlayerActive) {
                moveToNextPlayer(gameStatus)
            } else {
                moveToNextPhase(gameStatus)
            }
        } else {
            moveToNextPhase(gameStatus)
        }
    }

    private fun moveToNextPlayer(gameStatus: GameStatus) {
        gameStatus.turn.passPriority(gameStatus)

        if (!autocontinueChecker!!.canPerformAnyAction(gameStatus)) {
            moveToNextPhase(gameStatus)
        }
    }

    private fun moveToNextPhase(gameStatus: GameStatus) {
        if (name == CleanupPhase.Companion.CL) {
            gameStatus.turn.currentTurnPlayer = gameStatus.nonCurrentPlayer.name
            gameStatus.turn.increaseTurnNumber()
        }

        val nextPhase = getNextPhase(gameStatus)
        gameStatus.turn.currentPhase = nextPhase.name
        gameStatus.turn.currentPhaseActivePlayer = gameStatus.turn.currentTurnPlayer
        nextPhase.action(gameStatus)

        if (!autocontinueChecker!!.canPerformAnyAction(gameStatus)) {
            nextPhase.next(gameStatus)
        }
    }
}
